### ISSUE TITLE
Umap behaves differently from 0.4.6 to 0.5.0

### ISSUE THREAD

**Issue by swapnilsayansaha** [lmcinnes/umap#559]
I was using umap to perform dimensionality reduction on the entre 20 newsgroup dataset in sklearn.

In 0.4.6., when I used the cosine distance metric and fit_transform, umap returned the labels without error.

However, in 0.5.0, umap gives the following error: "Cannot use scipy.linalg.eigh for sparse A with k >= N. Use scipy.linalg.eigh(A.toarray()) or reduce k." (this absolutely is not supposed to happen for such a large dataset with 18k++ rows).

I used "init='random'" argument to solve the issue. The resulting matrix had a lot of NaN values.

The issue disappeared when i downgraded to 0.4.6. It does point out something is broken inside 0.5.0

**Comment by lmcinnes** [comment]
I'll see if I can take a look at what it going on there. Not much has changed around the region you specify the error derived from, so it seems odd.

**Comment by lmcinnes** [comment]
I am having trouble reproducing this. Currently I am simply running:

```python
import sklearn.datasets
import umap

news = sklearn.datasets.fetch_20newsgroups_vectorized(subset='all')
umap.UMAP(metric="cosine").fit_transform(news.data)
```
with the current versions of umap-learn and pynndescent installed. This does not present any errors for me. Is there some specific preprocessing that is required (using TFIDF for example)?

**Comment by swapnilsayansaha** [comment]
Thanks for looking into it.. Here's the preprocessing with count and tfidf vectorizers:

```
dataset = fetch_20newsgroups(subset = 'all',shuffle = True, random_state = 0,remove=('headers','footers'))
vectorizer = CountVectorizer(stop_words='english',min_df=3)
tfidf_transformer = TfidfTransformer()
data_feat_vec = vectorizer.fit_transform(dataset.data)
data_feat = tfidf_transformer.fit_transform(data_feat_vec)
```

Here's the arguments passed to UMAP:

```
r = [1,2,3,5,10,20,50,100,300]
Umap_euc = umap.UMAP(n_components=r[i], metric='euclidean').fit_transform(data_feat) #iterate r[i] through a for loop
```

**Comment by lmcinnes** [comment]
I'll try to look at this tomorrow. It is surprising to me that there is an issue here. Most likely it is related to disconnected components, but I'll have to see.

**Comment by lmcinnes** [comment]
It turns out this was a long standing corner case bug. What triggered it in 0.5 was introducing disconnection distances, which make the multi-component layout more likely. It was having some extra small components in combination with ``n_components=1`` that triggered the bug. That should be fixed now -- the actual change is small, so you can either install from master, or simply hand-patch your current installation if you wish.

**Comment by swapnilsayansaha** [comment]
Thanks it works fine now.
